waveorder/io/writer.py

```diff
@@ -0,0 +1,297 @@
+import numpy as np
+import zarr
+import os
+from numcodecs import Blosc
+
+"""
+Method:
+
+create zarr memorystore
+construct hierarchy there
+place arrays into this memory store
+
+at save time, create a target store that is zarr directoryStore
+use zarr.copy_store to copy memory store to directory store
+
+"""
+
+# todo: implement hierarchies / groups
+# todo: add checks for inconsistent data shapes
+# todo: change to persistent arrays?
+
+
+# initialize upfront -- with file path and dimensions -- use zeros and replace later
+#       initial dimensions are just (p, t, z, y, x) = (1, 1, 1, 512, 512) ==> (file) 2.0.0._._
+#       Initial chunks are important, must determine this first.
+# as data is reconstructed, replace zeros using same zarr reference.
+#
+
+
+class WaveorderWriter:
+    """
+    given stokes or physical data, construct a standard hierarchy in zarr for output
+        should conform to the ome-zarr standard as much as possible
+
+    """
+    __builder = None
+    save_dir = None
+    store_path = None
+    store = None
+
+    def __init__(self, save_dir: str, datatype: str):
+        """
+        datatype is one of "stokes" or "physical"
+        :param save_dir:
+        :param datatype:
+        """
+
+        self._check_is_dir(save_dir)
+
+        if datatype == 'stokes':
+            self.__builder = StokesZarr()
+        elif datatype == 'physical':
+            self.__builder = PhysicalZarr()
+        else:
+            raise NotImplementedError()
+
+    def _check_is_dir(self, path):
+        # Upon init, check to make sure save path is a directory,
+        # if not, create that directory
+        if os.path.isdir(path):
+            self.save_dir = path
+        else:
+            os.mkdir(path)
+            self.save_dir = path
+
+    def add_store(self, name: str):
+        # Add Zarr store to existing save directory,
+        # if store already exists, open that store
```

Comment:
I don't think the writer should read stores at all.  Maybe just check if it exists and asks to overwrite or not.

This raises a question of, what do we do when someone wants to load and modify an existing .zarr store?  This codebase is focused on reading mm tiffs (reader) and writing comp.micro reconstructions to zarr (writer) ... Maybe for post-compute modifications we leave that to the standard zarr library?

Reply:
I think that if reconstruction went poorly and you want to go back to a certain dataset to overwrite it, then being able to quickly switch stores is useful.  This function has since been replaced by "create_zarr()".

I do think we need some sort of safeguard to prevent the user from accidentally overwriting the data.  Maybe we can prompt the user to agree before continuing if an overwrite is possible.  There is definitely utility in being able to overwrite the data as you play around with reconstruction parameters.